### DISCUSSION TITLE
Error Response Standard

### DISCUSSION THREAD

**Discussion by arutr** [strapi/strapi#7371] [Legacy Topics]
Throughout the code, I'm seeing different ways of returning errors within endpoints:
https://github.com/strapi/strapi/blob/7113f7f11f391c571a317004b63b3daf5fc5e25e/packages/strapi-plugin-users-permissions/controllers/UsersPermissions.js#L20
https://github.com/strapi/strapi/blob/7113f7f11f391c571a317004b63b3daf5fc5e25e/packages/strapi-plugin-users-permissions/controllers/UsersPermissions.js#L31
https://github.com/strapi/strapi/blob/7b19532728785878450d82636238d65fe54938f1/packages/strapi-plugin-upload/controllers/upload/admin.js#L155

This makes error handling inconvenient on the front-end. Is there a standardised way of returning endpoint errors currently?

**Comment by derrickmehaffy** [comment]
Pinging @alexandrebodin to get his thoughts. Most of the ways done in the users-permissions plugin are old and need to be updated AFAIK.

The standard way that I am aware of is a `return ctx.badRequest('Your message here')` (and those options come from the boom middleware: https://github.com/strapi/strapi/blob/ad68fbb7b5d9a45dd8da4e96b036959b22fbc058/packages/strapi/lib/middlewares/boom/index.js#L12-L43

@lauriejim once we have a good answer for this, I'm going to add it to our to-do to document this as well, as it's a question I am seeing a lot lately.

**Reply by alexandrebodin** [reply]
FYI, 

Calling ctx.badRequest or throwing strapi.errors.badRequest will give your the same response. This is different just because you can throw a strapi.errors outside of controllers when you don't have access to the ctx.


What needs to be review is the error format. This is in the tech roadmap but not high priority compared to the rest of the roadmap :)

**Reply by derrickmehaffy** [reply]
In the backend @alexandrebodin I assume `strapi.errors.badRequest` === `ctx.badRequest` yeah? Still uses the same boom middleware?

**Reply by alexandrebodin** [reply]
Yep

**Reply by derrickmehaffy** [reply]
Great, thank you. @arutr I've added this to my to-do list to write some documentation about handling errors :smile:
